Use TCP keep alive to prevent connection freezing on slow connections

ben-palmer-sociomantic opened the pull request:
The mysql connection could indefinately freeze when in non-blocking mode
and on a bad internet connection. Set the TCP keep alive option to keep
the connection alive.

In addition the following other options are set:
Idle time in seconds after which to send keep alive probes - 3 seconds
Maximum number of keep alive probes before the connection is declared
dead and dropped - 3
Time in seconds between keep alive probes - 3

ben-palmer-sociomantic committed:
Use TCP keep alive to prevent connection freezing on slow connections

The mysql connection could indefinately freeze when in non-blocking mode
and on a bad internet connection. Set the TCP keep alive option to keep
the connection alive.

The options are set using the socket options struct.

ben-palmer-sociomantic commented:
Blocked until tested.

andreas-bok-sociomantic commented:
Could you change the base to **v5.4.x**?

ben-palmer-sociomantic commented:
> Could you change the base to v5.4.x?

Is it a bug fix though or a new feature? I was thinking a new feature.

andreas-bok-sociomantic commented:
> Is it a bug fix though or a new feature? I was thinking a new feature.

Ah true, v5.x.x is the right one then

andreas-bok-sociomantic commented:
Comparing with the source code in "classic" drizzle library, the KEEPALIVE option on the setsockopt is removed in the libdrizzle-redux fork. 
libdrizzle-redux  http://bazaar.launchpad.net/~drizzle-trunk/drizzle/7.2/view/head:/drizzled/plugin/listen_tcp.cc#L128

andreas-bok-sociomantic commented:
LGTM, just wondering if we should add the keep alive values as settable options.

gautam-kotian-sociomantic commented:
Typo in `indefinately` in commit message. LGTM otherwise.

ben-palmer-sociomantic commented:
> just wondering if we should add the keep alive values as settable options.

Yeah I think this is probably a good idea, so that will add four new options.

andreas-bok-sociomantic commented:
cool

andreas-bok-sociomantic commented:
Probably the best place to add it is on the drizzle_options_struct, https://github.com/sociomantic-tsunami/libdrizzle-redux/blob/v5.x.x/libdrizzle/structs.h#L154

andreas-bok-sociomantic committed:
Add struct to hold options for socket connection

The socket options include the socket timeout and the settings needed
when setting the TCP keep alive option.

andreas-bok-sociomantic committed:
Add methods to create and set socket options

andreas-bok-sociomantic committed:
Add use of socket options struct to unittest

andreas-bok-sociomantic commented:
Created a PR which allow the user to specify socket connection options, https://github.com/ben-palmer-sociomantic/libdrizzle-redux/pull/1

ben-palmer-sociomantic commented:
Updated with additional commits from @andreas-bok-sociomantic

ben-palmer-sociomantic commented:
Closing as this is replaced by https://github.com/sociomantic-tsunami/libdrizzle-redux/pull/114
